src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1390,28 +1390,38 @@ private bool ShouldCurrentContextualKeywordBeTreatedAsModifier()
 
             this.EatToken(); //move past contextual keyword
 
-            if (this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword)
+            if (parsingStatementNotDeclaration)
             {
-                this.EatToken(); // "partial" doesn't affect our decision, so look past it.
+                if (ScanType() != ScanTypeFlags.NotType)
```

Comment:
I don't understand why this is true. `ScanType()` seems to imply that it could return `GenericTypeOrMethod`, which would seem to imply that the following code could get through here: `async M<T>()`. However, `async` in that context could indicate a return type of `async`, instead of a modifier, could it not?